assets/js/phoenix.js

```diff
@@ -826,20 +827,26 @@ export class Socket {
    * Passing params to connect is deprecated; pass them in the Socket constructor instead:
    * `new Socket("/socket", {params: {user_id: userToken}})`.
    */
-  connect(params){
+  connect(params, callback){
     if(params){
       console && console.log("passing params to connect is deprecated. Instead pass :params to the Socket constructor")
       this.params = closure(params)
     }
-    if(this.conn){ return }
+    if(this.conn){ 
+      callback && callback()
+      return 
+    }
     this.closeWasClean = false
-    this.conn = new this.transport(this.endPointURL())
-    this.conn.binaryType = this.binaryType
-    this.conn.timeout    = this.longpollerTimeout
-    this.conn.onopen     = () => this.onConnOpen()
-    this.conn.onerror    = error => this.onConnError(error)
-    this.conn.onmessage  = event => this.onConnMessage(event)
-    this.conn.onclose    = event => this.onConnClose(event)
+    this.endPointURL().then(url => {
```

Comment:
Since a few ppl have this need, I think we should figure something out here.  Does anybody have an example app we can use to hone in on this solution/alternative solutions?

Reply:
I've published packages to test the patch

```json
    "@xfumihiro/absinthe-socket": "0.2.1-rc1",
    "@xfumihiro/absinthe-socket-apollo-link": "0.2.1-rc2",
    "@xfumihiro/phoenix": "1.4.0-rc1",
```
```js
createAbsintheSocketLink(
      AbsintheSocket.create(
        new PhoenixSocket(WS_URL, {
          reconnect: true,
          params: async () => {
            const accessToken = await getAccessToken()
            const refreshToken = await getRefreshToken()
            const token = accessToken || refreshToken
            return {
                jwt: token,
            }
          },
    }),
  ),
)
```

Reply:
So my first instinct is what is preventing a user from awaiting these tokens before stamping out a new `PhoenixSocket` instance?  That is the values are resolved before assigning to `params`... 

Note, I'm not implying this PR is unnecessary or anything.  Just want to understand, from your point of view, what Phoenix should handle and what we should depend on users to implement.

Reply:
as discussed in #3515 we need this patch in order for the connect method in PhoenixSocket to get the latest params (ex. tokens) asynchronously. again this isn't necessary for web where we can access **localStorage/cookies** synchronously but is critical for mobile where **AsyncStorage** can only be access asynchronously

Reply:
Yep the async access makes sense and is solved with async/await.  However, my curiosity is in making createAbsintheSocketLink async and awaiting the access token before calling `new PhoenixSocket`.  It sounds like you are saying it is a timing issue?  Ie asking for the token in createAbsintheSocketLink vs the connect method?

Reply:
not sure what you mean but the createAbsintheSocketLink is untouched. we only need the params (in PhoenixSocket constructor) to accept async functions.